File: src/algorithms/fscan.js
/**
 * F-SCAN Algorithm
 *
 * Splits the request queue into two sub-queues:
 *   - Active queue (F1): frozen snapshot of all pending requests when a scan begins.
 *   - Deferred queue (F2): requests that arrive while F1 is being processed (simulated here
 *     by splitting the input queue in half at the midpoint to illustrate the two-queue concept).
 *
 * The head sweeps through F1 using SCAN, then F2 becomes the new F1, and the process repeats.
 * This prevents arm starvation and arm "stickiness".
 *
 * Simulation note: Since all requests are known upfront, we split them into two equal halves
 * to demonstrate the two-pass F-SCAN behaviour.  In a live OS, F2 would accumulate dynamically.
 *
 * @param {number[]} requests  - Array of track numbers to service
 * @param {number}   headStart - Initial head position
 * @param {number}   diskSize  - Total number of tracks (default 200)
 * @param {string}   direction - Initial scan direction: 'right' | 'left'
 */
export function calculateFSCAN(requests, headStart, diskSize = 200, direction = 'right') {
    let currentHead = headStart;
    let totalSeek = 0;
    const steps = [];

    // Initial state
    steps.push({
        track: currentHead,
        seek: 0,
        totalSeek: 0,
        requestsLeft: [...requests],
        description: 'Start'
    });

    /**
     * Split into two queues:
     * F1 - first half (the "frozen active" queue)
     * F2 - second half  (the "deferred new arrivals" queue)
     */
    const mid = Math.ceil(requests.length / 2);
    const queues = [
        requests.slice(0, mid),  // F1
        requests.slice(mid)       // F2
    ];

    const processed = new Set();
    let currentDirection = direction;

    for (let q = 0; q < queues.length; q++) {
        const queue = queues[q];
        if (queue.length === 0) continue;

        const label = q === 0 ? 'F1 (Active)' : 'F2 (Deferred)';

        // Sort into left / right of current head for SCAN
        let left = queue.filter(r => r < currentHead).sort((a, b) => b - a);
        let right = queue.filter(r => r >= currentHead).sort((a, b) => a - b);

        let sequence = [];
        if (currentDirection === 'right') {
            sequence = [...right];
            if (left.length > 0) {
                sequence.push(diskSize - 1); // go to far end
                sequence = [...sequence, ...left];
            }
        } else {
            sequence = [...left];
            if (right.length > 0) {
                sequence.push(0); // go to near end
                sequence = [...sequence, ...right];
            }
        }

        for (let i = 0; i < sequence.length; i++) {
            const target = sequence[i];
            const distance = Math.abs(target - currentHead);

            if (distance >= 0) {
                totalSeek += distance;
                currentHead = target;

                const isRequest = queue.includes(target);
                if (isRequest) processed.add(target);

                steps.push({
                    track: target,
                    seek: distance,
                    totalSeek,
                    requestsLeft: requests.filter(r => !processed.has(r)),
                    description: isRequest
                        ? `Serve ${target} [${label}]`
                        : `Scan to End/Start [${label}]`
                });
            }
        }

        // Flip direction after each queue (F-SCAN uses SCAN internally)
        currentDirection = currentDirection === 'right' ? 'left' : 'right';
    }

    return { steps, totalSeek };
}

File: src/algorithms/fscan.c
#include "../scheduler/scheduler.h"
#include <stdlib.h>

int compare(const void *a, const void *b);

/*
 * FSCAN: Freeze-SCAN
 * Uses two queues. The current queue is "frozen" and serviced using SCAN.
 * New requests go into the second queue. Here we treat the entire input
 * as one frozen batch (since we don't have real-time arrival).
 * Essentially equivalent to a single SCAN pass over all requests.
 */
int fscan(int head, int arr[], int n, int direction, int disk_size) {
  int totalSeek = 0;
  int current = head;

  int sorted[256];
  for (int i = 0; i < n; i++)
    sorted[i] = arr[i];

  qsort(sorted, n, sizeof(int), compare);

  if (direction == 1) { /* RIGHT */
    for (int i = 0; i < n; i++) {
      if (sorted[i] >= head) {
        totalSeek += abs(current - sorted[i]);
        current = sorted[i];
      }
    }

    totalSeek += abs(current - (disk_size - 1));
    current = disk_size - 1;

    for (int i = n - 1; i >= 0; i--) {
      if (sorted[i] < head) {
        totalSeek += abs(current - sorted[i]);
        current = sorted[i];
      }
    }
  } else { /* LEFT */
    for (int i = n - 1; i >= 0; i--) {
      if (sorted[i] <= head) {
        totalSeek += abs(current - sorted[i]);
        current = sorted[i];
      }
    }

    totalSeek += abs(current - 0);
    current = 0;

    for (int i = 0; i < n; i++) {
      if (sorted[i] > head) {
        totalSeek += abs(current - sorted[i]);
        current = sorted[i];
      }
    }
  }

  return totalSeek;
}
